keps/sig-api-machinery/20190425-structural-openapi.md

```diff
@@ -0,0 +1,750 @@
+        
+        // x-kubernetes-embedded-resource defines that the value is an 
+        // embedded Kubernetes runtime.Object, with TypeMeta and 
+        // ObjectMeta. The type must be object. It is allowed to further
+        // restrict the embedded object. Both ObjectMeta and TypeMeta
+        // are validated automatically. x-kubernetes-preserve-unknown-fields
+        // must be true.
+        XEmbeddedResource bool
+        
+        // x-kubernetes-int-or-string specifies that this value is 
+        // either an integer or a string. It is allowed to 
+        XIntOrString bool
+}
+
+type Generic struct {
+	// type is specifies the type of a value.
+	// It can be object, array, number, integer, boolean, string.
+	// It is optional only if x-kubernetes-preserve-unknown-fields 
+	// or x-kubernetes-int-or-string specifies is true.
+	Type        string
+	AdditionalProperties *StructuralSchemaOrBool
+	
+	Description string
+        Nullable    bool
+	Default     *JSON
+        ReadOnly    bool
+}
+
+type ValueValidation struct {
+	Format               string
+	Maximum              *float64
+        ExclusiveMaximum     bool
+        Minimum              *float64
+        ExclusiveMinimum     bool
+        MaxLength            *int64
+        MinLength            *int64
+        Pattern              string
+        MaxItems             *int64
+        MinItems             *int64
+        UniqueItems          bool
+        MultipleOf           *float64
+        Enum                 []JSON
+        MaxProperties        *int64
+        MinProperties        *int64
+        Required             []string
+        AllOf                []NestedValueValidation
+        OneOf                []NestedValueValidation
+        AnyOf                []NestedValueValidation
+        Not                  *NestedValueValidation
+}
+
+type NestedValueValidation struct {
+	ValueValidation
+
+	Items      *NestedValueValidation
+        Properties map[string]NestedValueValidation
+        
+        // Anything set in the following will make the scheme 
+        // non-structural, with the exception of these two patterns if 
+        // x-kubernetes-int-or-string is true:
+        //
+        // 1) anyOf:
+        //    - type: integer
+        //    - type: string
+        // 2) allOf:
+        //    - anyOf:
+        //      - type: integer
+        //      - type: string
+        //    - <NestedValueValidation>
+        ForbiddenGenerics Generic `"json:inline"`
+        ForbiddenExtensions Extensions `"json:inline"`
+}
+```
+
+During CRD validation, we verify in addition to existing validations:
+ 
+1. **structurality:** that both `ForbiddenGenerics` and `ForbiddenExtensions` only have zero values, with the two exceptions of (4).
+2. **completeness:** that the schema is _structurally complete_
+3. **RawExtension:** that for every schema with `x-kubernetes-embedded-resource: true` `x-kubernetes-preserve-unknown-fields: true` and `type: object` are set
```

Comment:
@liggitt do we want to enforce `x-kubernetes-preserve-unknown-fields: true` and `type: object` as written here? I tend to do this as it restricts the special case even more.

Reply:
sure